### DISCUSSION TITLE
Feature request and use case example to allow `matrix` in `if`s

### DISCUSSION THREAD

**Discussion by MOZGIII** [community/community#25622] [Actions]
<p>Currently, using <code>matrix</code> at <code>jobs.&lt;job&gt;.if</code> is an error.</p>
<p>We’re building an E2E test suite at <a href="https://github.com/timberio/vector" rel="nofollow noopener">https://github.com/timberio/vector</a>, and want to use matrices and labels to conditionally skip some of the test cases.</p>
<p>Here’s an example of what we’d like to have:</p>
<pre><code class="lang-yaml">name: E2E Suite

on:
  push: {}
  pull_request: {}
  schedule:
    - cron: "0 4 * * *"

jobs:
  test-e2e-kubernetes:
    name: K8s ${{ matrix.kubernetes_version.version }} / ${{ matrix.container_runtime.name }}
    runs-on: ubuntu-latest
    strategy:
      matrix:
        kubernetes_version:
          - version: "v1.18.6"
            is_essential: true
          - version: "v1.17.9"
        container_runtime:
          - name: docker
            is_essential: true
          - name: containerd
    if: |
          (
            matrix.kubernetes_version.is_essential == true &amp;&amp;
            matrix.container_runtime.is_essential == true
          ) || (
            github.event.pull_request == null ||
            contains(github.event.pull_request.labels.*.name, 'ci-condition: k8s e2e all targets')
          )
    steps: ...
</code></pre>
<p>This is a simplified example, to demonstrate the point.</p>
<p>The idea is we’d like to have a way to only run the full test suite in a nightly build, or if the PR has a certain label (<code>ci-condition: k8s e2e all targets</code> in our case). We can do that, but we also want our job definition to be compact - we don’t want to duplicate it.</p>
<p>The listing above doesn’t work, because currently we can’t use <code>matrix.kubernetes_version.is_essential</code> and <code>matrix.container_runtime.is_essential</code> in the <code>if:</code> statement.</p>
<p>We’d like to have the access to <code>matrix</code> at the <code>if</code> level.</p>
<p>We’ve also considered using <code>matrix.exclude</code>, however that doesn’t allow us to dynamically condition those exclusions based on <code>contains(github.event.issue.labels.*.name, 'ci-condition: k8s e2e all targets') || github.event_name == 'schedule'</code> condition.</p>
<p>UPD 1:</p>
<p>Another notable issue we encountered as a consequence of the inability to use <code>matrix</code> at the job-level <code>if:</code> clause is the inability to alter the job execution result beyond <code>success</code> / <code>failure</code>. We attempted to move the logic into the step-level <code>if:</code> statement, however doing that required us to have a third outcome state for our test - something like  <code>skipped</code> or <code>cancelled</code> - to indicate that the test didn’t show neither positive nor negative outcome. How can there be one if the test itself was effectively skipped? I remember there was a way to switch the <code>job</code> into a <code>cancelled</code> state, however, it seems it’s not available anymore.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/mozgiii">@mozgiii</a>,</p>
<p>Glad to see you in Github Community Forum!</p>
<p>Currently <code>matrix</code> is NOT supported in job <code>if</code> expression, it has been already raised as an feature_request, please check similar <a href="https://github.community/t/bug-trouble-with-if-conditional-expression-groupings/16563">ticket</a> for the details.</p>
<p>Actually  i’m a little confused with your if expression, since you use <code>||</code> there, there could be multiple situations with different events. It’s recommended to move push event to a new yaml.<br>
Could you please let me know for pull_request(or schedule) event, why to evaluate matrix value?</p>
<p>It’s not supported to change the step outcome, but cancel the whole job, you can try to use andymckay/cancel-action@0.2</p>
<p>Thanks.</p>

**Comment by MOZGIII** [comment]
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a>,</p>
<p>There was a bug in the if condition, the correct version would be</p>
<pre><code class="lang-auto">if: |
  (
    matrix.kubernetes_version.is_essential &amp;&amp;
    matrix.container_runtime.is_essential
  ) || (
    github.event.pull_request == null ||
    contains(github.event.pull_request.labels.*.name, 'ci-condition: k8s e2e all targets')
  )
</code></pre>
<p>Notice there’s a different condition at the second part of the <code>||</code>.</p>
<p>I think this is a good way to explain the logic we’re trying to achieve:</p>
<ul>
<li>if we’re executing at a PR with no labels - run only essential tests</li>
<li>if we’re executing at a PR with <code>ci-condition: k8s e2e all targets</code> label set - run all tests</li>
<li>if we’re executing at a non-PR (push to <code>master</code> or scheduled run on <code>master</code>) - run all tests</li>
</ul>
<p>In that sense - we first check if the test is essential - and if it is we always run the said test - otherwise check the context we’re executing at, and <em>sometimes</em> run the test even if it’s not essential.</p>
<p><code>github.event.pull_request == null</code> tells us if we’re not in a PR context - then we always want to run all tests<br>
<code>contains(github.event.pull_request.labels.*.name, 'ci-condition: k8s e2e all targets')</code> - if we’re at a PR - only run if there’s a certain label</p>
<p>I’ve looked into <code>andymckay/cancel-action@0.2</code> - however it cancels the whole workflow. We’d need to cancel just one particular job from the workflow.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/mozgiii">@mozgiii</a>,</p>
<p>Thanks for your reply!</p>
<p>Github support dynamic matrix value now. Based on your logic, you can add a <code>prejob</code>  and set the matrix value accordingly(events&amp;pr labels) , transfer the matrix value to your test job. In this case, you don’t need to cancel the redundant jobs.<br>
Code sample as below(please change the matrix value for yours):</p>
<pre><code class="lang-auto">on:
  push: 
  pull_request: 
  schedule:
    - cron: "*/6 * * * *"

jobs:
  prejob:
    runs-on: ubuntu-latest
    outputs:
      matrix: ${{ steps.set-matrix1.outputs.matrix }}${{ steps.set-matrix2.outputs.matrix }}  #get the matrix value and transfer to next job
    steps:
      - name: set matrix 1   # set the matrix value for push, schedule event and pr event with expected label
        id: set-matrix1
        if: (github.event_name=='push') || (github.event_name=='schedule') || (github.event_name=='pull_request' &amp;&amp; contains(github.event.pull_request.labels.*.name, 'ci-condition:k8s e2e all targets'))
        run: |
          echo "::set-output name=matrix::{\"include\":[{\"project\":\"foo\",\"config\":\"Debug\"},{\"project\":\"bar\",\"config\":\"Release\"}]}"
      - name: set matrix 2  # set the matrix value for pull_request event without expetected label
        id: set-matrix2
        if: github.event_name=='pull_request' &amp;&amp; !contains(github.event.pull_request.labels.*.name, 'ci-condition:k8s e2e all targets')
        run: |
          echo "::set-output name=matrix::{\"include\":[{\"test\":\"value1\",\"config\":\"Debug1\"},{\"test\":\"value2\",\"config\":\"Release1\"}]}"
            

  test-e2e-kubernetes:
    needs: [prejob]
    runs-on: ubuntu-latest
    strategy:
      matrix: ${{fromJson(needs.prejob.outputs.matrix)}}   # get the matrix value

    steps: 
      - name: Dump matrix context
        env:
          MATRIX_CONTEXT: ${{ toJson(matrix) }}
        run: echo "$MATRIX_CONTEXT"
</code></pre>
<p>Confirmed it works on my side. For exmaple:<br>
push event workflow: <a href="https://github.com/weide-zhou/ticket13/actions/runs/197233132" rel="nofollow noopener">https://github.com/weide-zhou/ticket13/actions/runs/197233132</a><br>
pull request event without expected labels: <a href="https://github.com/weide-zhou/ticket13/actions/runs/197237243" rel="nofollow noopener">https://github.com/weide-zhou/ticket13/actions/runs/197237243</a><br>
pull request with expected label: <a href="https://github.com/weide-zhou/ticket13/actions/runs/197234906" rel="nofollow noopener">https://github.com/weide-zhou/ticket13/actions/runs/197234906</a></p>
<p>Please refer to the <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#example-6" rel="nofollow noopener">doc</a> for more details.</p>
<p>Thanks</p>

**Comment by MOZGIII** [comment]
<p>Wow, amazing! A very nice solution! I didn’t have a clue to look into that direction. Github definitely should explicitly add a hint to the matrix doc that matrix can be assigned that way.</p>
<p>Thanks a lot, I’ll try this!</p>
<p>UPD: Works like a charm, thanks again!</p>

**Comment by arcivanov** [comment]
<p>This still doesn’t appear to work with <code>if</code> condition using <code>matrix</code> on a <code>job</code> as opposed to a step.</p>
